gerritlab/global_vars.py

```diff
@@ -27,28 +34,122 @@ def load_config(remote, repo):
     root_dir = repo.git.rev_parse("--show-toplevel")
-    config.read(os.path.join(root_dir, ".gitreview"))
-    configs = config[remote]
-    host = configs["host"]
-    project_id = configs["project_id"]
-    if "GITLAB_PRIVATE_TOKEN" in os.environ:
-        private_token = os.environ["GITLAB_PRIVATE_TOKEN"]
-    elif "private_token" in configs:
-        private_token = configs["private_token"]
-    else:
-        private_token = repo.config_reader().get_value("gerritlab", "private-token")
+    git_config = repo.config_reader()
+
+    username = git_config.get_value("user", "name")
+    email = git_config.get_value("user", "email")
+
+    gitreview = configparser.ConfigParser()
+    gitreview.read(os.path.join(root_dir, ".gitreview"))
+    gitreview_config = gitreview[remote]
+
+    host = gitreview_config["host"]
+    project_id = gitreview_config["project_id"]
+
+    private_token = _get_private_token(host, git_config, gitreview_remote)
+
     # Optional configs.
-    if "target_branch" in configs:
-        global_target_branch = configs["target_branch"]
-    if "remove_source_branch" in configs:
-        remove_source_branch = configs.getboolean("remove_source_branch")
+    if "target_branch" in gitreview_config:
+        global_target_branch = gitreview_config["target_branch"]
+    if "remove_source_branch" in gitreview_config:
+        remove_source_branch = gitreview_config.getboolean("remove_source_branch")
     mr_url = "{}/api/v4/projects/{}/merge_requests".format(host, project_id)
     pipeline_url = "{}/api/v4/projects/{}/pipeline".format(host, project_id)
     pipelines_url = "{}/api/v4/projects/{}/pipelines".format(host, project_id)
     session = requests.session()
     session.headers.update({"PRIVATE-TOKEN": private_token})
+
+
+def _get_private_token(host: str,
+                       git_config: GitConfigParser,
+                       gitreview_config: configparser.SectionProxy):
+    """
+    Get the private token from one of several sources in the following order of
+    preference:
+
+    1. The GITLAB_PRIVATE_TOKEN environment variable.
+    2. git credential storage
+    3. git config
+    3. (DEPRECATED) The .gitreview file
+
+    Kind of a messy function, but it hides this complexity from the
+    `load_config` function.
+    """
+    if "GITLAB_PRIVATE_TOKEN" in os.environ:
+        return os.environ["GITLAB_PRIVATE_TOKEN"]
+
+    # Try to get the private token from git credentials.
+    private_token = GitCredentialStore(host).get_token()
+    if private_token is not None:
+        return private_token
+
+    # Try to get the private token from git config.
+    try:
+        return git_config.get_value("gerritlab", "private-token")
+    except (configparser.NoSectionError, configparser.NoOptionError):
+        pass
+
+    # DEPRECATED: Try to get the private token from .gitreview file.
+    if "private_token" in gitreview_config:
+        warn("[DEPRECATED] Using private_token from .gitreview file is deprecated.")
```

Comment:
Can we throw an error instead if this is to be deprecated?

Reply:
Done! Hard deprecated now.